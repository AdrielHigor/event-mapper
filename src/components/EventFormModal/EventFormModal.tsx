import React, { useState } from "react";
import "./EventFormModal.css";
import ModalContainer from "../ModalContainer/ModalContainer";
import FormInput from "../FormTextInput/FormInput";
import {
  DynamicObject,
  IEventForm,
  IEventFormValidation,
} from "../../utils/interfaces/base";

interface IEventFormModal {
  modalIsOpen: boolean;
  closeModal: (
    event: React.MouseEvent<Element, MouseEvent> | React.KeyboardEvent<Element>
  ) => void;
  customClassName?: string;
  title: string;
  description: string;
  saveButtonLabel: string;
  cancelButtonLabel: string;
  form?: IEventForm;
  requiredFields?: Array<string>;
  onFormChange: (e: IEventForm) => void;
  onSave: () => void;
  onCancel: () => void;
}

const EventFormModal = ({
  modalIsOpen,
  closeModal,
  customClassName,
  title,
  description,
  saveButtonLabel,
  cancelButtonLabel,
  form,
  requiredFields,
  onFormChange,
  onSave,
  onCancel,
}: IEventFormModal) => {
  const [formValidFields, setFormValidFields] = useState<IEventFormValidation>({
    name: true,
    description: true,
    startDateTime: true,
    endDateTime: true,
  });

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const updatedObj: DynamicObject = { ...form };
    updatedObj[e.target.id] = e.target.value;
    onFormChange(updatedObj);
  };

  const handleSave = (validatedFields: DynamicObject) => {
    let field;

    for (field in validatedFields) {
      if (validatedFields[field] === false) return;
    }

    onSave();
  };

  const handleValidation = () => {
    const validatedFields: DynamicObject = { ...formValidFields };
    const formObj: DynamicObject = { ...form };
    if (requiredFields) {
      requiredFields.forEach((field) => {
        if (!formObj[field] || formObj[field] === "") {
          validatedFields[field] = false;
        } else {
          validatedFields[field] = true;
        }
      });

      setFormValidFields(validatedFields);
    }

    handleSave(validatedFields);
  };

  return (
    <ModalContainer
      modalIsOpen={modalIsOpen}
      closeModal={closeModal}
      customClassName={customClassName}
    >
      <div className="event-form">
        <div className="modal-header">
          <h2>{title}</h2>
          <p className="description">{description}</p>
        </div>
        <div className="form">
          <FormInput
            onChange={handleChange}
            id="name"
            label="Nome do evento *"
            isValid={formValidFields["name"]}
          />
          <FormInput
            onChange={handleChange}
            id="description"
            label="Descrição do evento *"
            isValid={formValidFields["description"]}
          />
          <FormInput
            onChange={handleChange}
            id="startDateTime"
            label="Data de início do evento"
            inputType="datetime-local"
            isValid={formValidFields["startDateTime"]}
          />
          <FormInput
            onChange={handleChange}
            id="endDateTime"
            label="Data de fim do evento"
            inputType="datetime-local"
            isValid={formValidFields["endDateTime"]}
          />
        </div>
        <div className="options">
          <button className="confirm-button" onClick={handleValidation}>
            {saveButtonLabel}
          </button>
          <button className="cancel-button" onClick={onCancel}>
            {cancelButtonLabel}
          </button>
        </div>
      </div>
    </ModalContainer>
  );
};

export default EventFormModal;
