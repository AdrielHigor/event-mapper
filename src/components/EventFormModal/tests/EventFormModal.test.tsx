import renderer from "react-test-renderer";
import EventFormModal, { IEventFormModal } from "../EventFormModal";

describe("EventFormModal", () => {
  const defaultProps: IEventFormModal = {
    modalIsOpen: true,
    closeModal: jest.fn(),
    title: "Test Title",
    description: "Test Description",
    saveButtonLabel: "Save",
    cancelButtonLabel: "Cancel",
    onFormChange: jest.fn(),
    onSave: jest.fn(),
    onCancel: jest.fn(),
  };

  const defaultForm = {
    name: "Test Name",
    description: "Test Description",
    startDateTime: "2023-05-01T12:00",
    endDateTime: "2023-05-01T13:00",
  };

  const wrapper = (options = defaultProps) => {
    return renderer.create(<EventFormModal {...options} />);
  };

  it("renders correctly", () => {
    const component = wrapper();
    expect(component).toMatchSnapshot();
  });

  it("calls onCancel when the cancel button is clicked", () => {
    const component = wrapper();
    const cancelButton = component.root.findByProps({
      className: "cancel-button",
    });
    cancelButton.props.onClick();
    expect(defaultProps.onCancel).toHaveBeenCalledTimes(1);
  });

  it("calls onSave when the save button is clicked and all fields are valid", () => {
    const onSaveMock = jest.fn();
    const component = wrapper({
      ...defaultProps,
      form: defaultForm,
      requiredFields: ["name", "description", "startDateTime", "endDateTime"],
      onSave: onSaveMock,
    });

    const saveButton = component.root.findByProps({
      className: "confirm-button",
    });
    saveButton.props.onClick();
    expect(onSaveMock).toHaveBeenCalled();
  });

  it("should call handleSave if no requiredFields", () => {
    const onSaveMock = jest.fn();
    const component = wrapper({
      ...defaultProps,
      form: defaultForm,
      onSave: onSaveMock,
    });

    const saveButton = component.root.findByProps({
      className: "confirm-button",
    });
    saveButton.props.onClick();
    expect(onSaveMock).toHaveBeenCalled();
  });

  it(`doesn't call onSave if button is clicked and at least one required field is empty`, () => {
    const onSaveMock = jest.fn();
    const form = {
      name: "Test Name",
      description: "",
      startDateTime: "2023-05-01T12:00",
      endDateTime: "2023-05-01T13:00",
    };

    const component = wrapper({
      ...defaultProps,
      form,
      requiredFields: ["name", "description", "startDateTime", "endDateTime"],
      onSave: onSaveMock,
    });

    const saveButton = component.root.findByProps({
      className: "confirm-button",
    });
    saveButton.props.onClick();
    expect(onSaveMock).not.toHaveBeenCalled();
  });

  it("should call onFormChange when input value changes", () => {
    const onFormChangeMock = jest.fn();
    const form = {
      name: "",
      description: "",
      startDateTime: "",
      endDateTime: "",
    };

    const component = wrapper({
      ...defaultProps,
      form,
      onFormChange: onFormChangeMock,
    });

    const input = component.root.findByProps({ id: "name" });
    input.props.onChange({ target: { id: "name", value: "New event name" } });

    expect(onFormChangeMock).toHaveBeenCalledWith({
      name: "New event name",
      description: "",
      startDateTime: "",
      endDateTime: "",
    });
  });
});
