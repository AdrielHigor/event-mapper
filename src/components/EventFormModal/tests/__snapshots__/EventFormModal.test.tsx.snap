// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventFormModal renders correctly 1`] = `
<div>
  <div
    className="event-form"
  >
    <div
      className="modal-header"
    >
      <h2>
        Test Title
      </h2>
      <p
        className="description"
      >
        Test Description
      </p>
    </div>
    <div
      className="form"
    >
      <div
        className="input-container"
      >
        <label
          className="form-label false"
          htmlFor="name"
        >
          Nome do evento *
        </label>
        <input
          className="form-input false"
          id="name"
          onChange={[Function]}
        />
      </div>
      <div
        className="input-container"
      >
        <label
          className="form-label false"
          htmlFor="description"
        >
          Descrição do evento *
        </label>
        <input
          className="form-input false"
          id="description"
          onChange={[Function]}
        />
      </div>
      <div
        className="input-container"
      >
        <label
          className="form-label false"
          htmlFor="startDateTime"
        >
          Data de início do evento
        </label>
        <input
          className="form-input false"
          id="startDateTime"
          onChange={[Function]}
          type="datetime-local"
        />
      </div>
      <div
        className="input-container"
      >
        <label
          className="form-label false"
          htmlFor="endDateTime"
        >
          Data de fim do evento
        </label>
        <input
          className="form-input false"
          id="endDateTime"
          onChange={[Function]}
          type="datetime-local"
        />
      </div>
    </div>
    <div
      className="options"
    >
      <button
        className="confirm-button"
        onClick={[Function]}
      >
        Save
      </button>
      <button
        className="cancel-button"
        onClick={[MockFunction]}
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
